### PR TITLE
Exclude frameworks option

### LINKED ISSUE #30: Add option to exclude frameworks from obfuscation
Add option to exclude all or selected frameworks from obfuscation

### PR DESCRIPTION
**Pull request by kjamroz-bt**
Closes #30 
Requires getup_long (#29) to be merged first.

### CODE REVIEW

File: MachObfuscator/DependencyAnalysis/ObfuscationPaths+Building.swift [hunk 1]
```diff
@@ -1,13 +1,14 @@
 import Foundation
 
 extension ObfuscationPaths {
-    static func forAllExecutablesWithDependencies(inDirectory dir: URL, fileRepository: FileRepository = FileManager.default, dependencyNodeLoader: DependencyNodeLoader) -> ObfuscationPaths {
+    static func forAllExecutablesWithDependencies(inDirectory dir: URL, fileRepository: FileRepository = FileManager.default, dependencyNodeLoader: DependencyNodeLoader, obfuscableFilesFilter: ObfuscableFilesFilter) -> ObfuscationPaths {
         var paths = ObfuscationPaths()
-        // TODO: create better exclusion mechanism (with custom excludes and includes)
+        
+        // TODO: create better exclusion mechanism (with custom excludes and includes)
```
Comment (kam800) [inline MachObfuscator/DependencyAnalysis/ObfuscationPaths+Building.swift]: This TODO can probably be removed now :)

File: MachObfuscator/Options/Options.swift [hunk 2]
```diff
@@ -28,26 +36,66 @@ extension Options {
...
+        };
+        
+        let longopts: [option] = [
+            option(name: Options.newCCharPtrFromStaticString("help"),      has_arg: no_argument,       flag: nil, val: OptLongChars.help),
+            option(name: Options.newCCharPtrFromStaticString("verbose"),   has_arg: no_argument,       flag: nil, val: OptLongChars.verbose),
+            option(name: Options.newCCharPtrFromStaticString("methtype"),      has_arg: no_argument, flag: nil, val: OptLongChars.methTypeObfuscation),
+            option(name: Options.newCCharPtrFromStaticString("machoview-doom"),      has_arg: no_argument, flag: nil, val: OptLongChars.machOViewDoom),
+            option(name: Options.newCCharPtrFromStaticString("swift-reflection"),      has_arg: no_argument, flag: nil, val: OptLongCases.swiftReflection.rawValue),
+            option(name: Options.newCCharPtrFromStaticString("no-framework"),      has_arg: required_argument, flag: nil, val: OptLongCases.noFramework.rawValue),
+            option(name: Options.newCCharPtrFromStaticString("no-frameworks"),      has_arg: no_argument, flag: nil, val: OptLongCases.noFrameworks.rawValue),
```
Comment (kam800) [inline MachObfuscator/Options/Options.swift]: What do you think about changing option key to `skip-all-frameworks`?

File: MachObfuscator/Options/Options.swift [hunk 3]
```diff
@@ -28,26 +36,66 @@ extension Options {
...
+            case noFrameworks
+        };
+        
+        let longopts: [option] = [
+            option(name: Options.newCCharPtrFromStaticString("help"),      has_arg: no_argument,       flag: nil, val: OptLongChars.help),
+            option(name: Options.newCCharPtrFromStaticString("verbose"),   has_arg: no_argument,       flag: nil, val: OptLongChars.verbose),
+            option(name: Options.newCCharPtrFromStaticString("methtype"),      has_arg: no_argument, flag: nil, val: OptLongChars.methTypeObfuscation),
+            option(name: Options.newCCharPtrFromStaticString("machoview-doom"),      has_arg: no_argument, flag: nil, val: OptLongChars.machOViewDoom),
+            option(name: Options.newCCharPtrFromStaticString("swift-reflection"),      has_arg: no_argument, flag: nil, val: OptLongCases.swiftReflection.rawValue),
+            option(name: Options.newCCharPtrFromStaticString("no-framework"),      has_arg: required_argument, flag: nil, val: OptLongCases.noFramework.rawValue),
```
Comment (kam800) [inline MachObfuscator/Options/Options.swift]: What do you think about changing option name to `skip-framework`?

File: MachObfuscator/DependencyAnalysis/ObfuscableFilesFilter.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }
+    
+    static func notFramework(framework : String) -> ObfuscableFilesFilter {
+        return ObfuscableFilesFilter { url in
+            !url.lastPathComponent.starts(with: framework)
+        }
+    }
+    
+    static func notFrameworks() -> ObfuscableFilesFilter {
```
Comment (kam800) [inline MachObfuscator/DependencyAnalysis/ObfuscableFilesFilter.swift]: What do you think about `skipAllFrameworks` and also changing the remaining names: `skipFramework` and `skipSwiftLibraries`?

File: MachObfuscator/Options/Options.swift (hunk 2 above)
Comment (kam800) [inline MachObfuscator/Options/Options.swift]: What do you think about changing option name to `skip-all-frameworks`?

File: MachObfuscator/Options/Options.swift (hunk 3 above)
Comment (kam800) [inline MachObfuscator/Options/Options.swift]: What do you think about changing option name to `skip-framework`?

File: MachObfuscator/DependencyAnalysis/ObfuscationPaths+Building.swift (hunk 1 above)
Comment (kam800) [inline MachObfuscator/DependencyAnalysis/ObfuscationPaths+Building.swift]: This TODO can probably be removed now :)

File: MachObfuscator/DependencyAnalysis/ObfuscableFilesFilter.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    
+    static func onlyFiles(in obfuscableDirectory: URL) -> ObfuscableFilesFilter {
+        return ObfuscableFilesFilter { url in
+            obfuscableDirectory.contains(url)
+        }
+    }
+    
+    static func notFramework(framework : String) -> ObfuscableFilesFilter {
+        return ObfuscableFilesFilter { url in
+            !url.lastPathComponent.starts(with: framework)
```
Comment (kam800) [inline MachObfuscator/DependencyAnalysis/ObfuscableFilesFilter.swift]: What do you think about changing this a little bit?
```
let frameworkComponent = framework + ".framework"
return ObfuscableFilesFilter { url in
    !url.components.contains(with: frameworkComponent)
}
```

**Comment by kam800** [comment]
Could you please install and run swiftformat? – `brew install swiftformat` and then just build the project in Xcode.

File: MachObfuscator/DependencyAnalysis/ObfuscableFilesFilter.swift (hunk 4 above)
Comment (kjamroz-bt) [reply MachObfuscator/DependencyAnalysis/ObfuscableFilesFilter.swift]: Done.

File: MachObfuscator/Options/Options.swift (hunk 2 above)
Comment (kjamroz-bt) [reply MachObfuscator/Options/Options.swift]: done

File: MachObfuscator/Options/Options.swift (hunk 3 above)
Comment (kjamroz-bt) [reply MachObfuscator/Options/Options.swift]: done

File: MachObfuscator/DependencyAnalysis/ObfuscationPaths+Building.swift (hunk 1 above)
Comment (kjamroz-bt) [reply MachObfuscator/DependencyAnalysis/ObfuscationPaths+Building.swift]: done

File: MachObfuscator/DependencyAnalysis/ObfuscableFilesFilter.swift (hunk 5 above)
Comment (kjamroz-bt) [reply MachObfuscator/DependencyAnalysis/ObfuscableFilesFilter.swift]: ok, changed

**Pull request merged**